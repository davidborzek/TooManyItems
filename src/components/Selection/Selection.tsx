import {
  Modal,
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native';

import { Input } from 'react-native-elements';
import { Ionicons } from '@expo/vector-icons';
import React from 'react';

export type SelectionItem<T> = {
  label: string;
  value: T;
};

type Props<T> = {
  items: SelectionItem<T>[];
  value: T;
  title?: string;
  placeholder?: string;
  label?: string;
  visible?: boolean;
  setOpen: (state: boolean) => void;
  onChange: (value: T) => void;
  EmptyStateComponent?: React.ReactElement | null;
};

/**
 * Ein wiederverwendbarer Component welcher Ähnlich wie ein Dropdown Menü funktioniert.
 * Die Auswahlmöglichkeiten werden am unteren Bildschirmrand angezeigt.
 */
export default function Selection<T>({
  items,
  visible,
  value,
  title,
  label,
  placeholder,
  setOpen,
  onChange,
  EmptyStateComponent,
}: Props<T>) {
  const renderItems = () => {
    return items.map((item, index) => (
      <TouchableOpacity
        key={index}
        onPress={() => {
          onChange(item.value);
          setOpen(false);
        }}
      >
        <Text style={styles.item}>{item.label}</Text>
      </TouchableOpacity>
    ));
  };

  return (
    <View>
      <Modal
        visible={visible}
        onRequestClose={() => setOpen(false)}
        animationType="fade"
        transparent
        statusBarTranslucent
      >
        <View style={styles.modal}>
          <TouchableWithoutFeedback onPress={() => setOpen(false)}>
            <View style={styles.overlay} />
          </TouchableWithoutFeedback>
          <View style={styles.modalContent}>
            <View style={styles.header}>
              <Text style={styles.heading}>{title}</Text>
              <TouchableOpacity onPress={() => setOpen(false)}>
                <Ionicons name={'close-outline'} size={32} />
              </TouchableOpacity>
            </View>
            {items.length === 0 && EmptyStateComponent ? (
              EmptyStateComponent
            ) : (
              <ScrollView>{renderItems()}</ScrollView>
            )}
          </View>
        </View>
      </Modal>
      <TouchableOpacity onPress={() => setOpen(true)}>
        <Input
          editable={false}
          label={label}
          placeholder={placeholder}
          autoCompleteType=""
          value={items.find((i) => i.value === value)?.label}
          enablesReturnKeyAutomatically
        />
      </TouchableOpacity>
    </View>
  );
}

const styles = StyleSheet.create({
  header: {
    alignItems: 'center',
    flexDirection: 'row',
    justifyContent: 'space-between',
    padding: 10,
  },
  heading: {
    fontSize: 16,
    fontWeight: '600',
  },
  item: {
    borderBottomColor: '#e1e1e1',
    borderBottomWidth: 1,
    padding: 16,
  },
  modal: {
    backgroundColor: 'rgba(0,0,0,0.5)',
    flex: 1,
    justifyContent: 'flex-end',
  },
  modalContent: { backgroundColor: 'white', height: '50%' },
  overlay: { flex: 1 },
});
