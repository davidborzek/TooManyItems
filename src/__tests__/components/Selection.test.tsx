import Selection, { SelectionItem } from '../../components/Selection/Selection';

import React from 'react';
import { Text } from 'react-native';
import renderer from 'react-test-renderer';

const items: SelectionItem<string>[] = [
  {
    label: 'Some label',
    value: 'Some value',
  },
];

it('Selection renders correctly with required props, no items and closed', () => {
  const tree = renderer
    .create(
      <Selection items={[]} onChange={() => ''} setOpen={() => ''} value="" />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Selection renders correctly with required props, with items and closed', () => {
  const tree = renderer
    .create(
      <Selection
        items={items}
        onChange={() => ''}
        setOpen={() => ''}
        value=""
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Selection renders correctly with required props, with items and opened', () => {
  const tree = renderer
    .create(
      <Selection
        items={items}
        onChange={() => ''}
        setOpen={() => ''}
        value=""
        visible
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Selection renders correctly with full props and opened', () => {
  const tree = renderer
    .create(
      <Selection
        items={items}
        onChange={() => ''}
        setOpen={() => ''}
        value=""
        visible
        EmptyStateComponent={<Text>Empty</Text>}
        label="Some label"
        placeholder="Some placeholder"
        title="Some title"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
