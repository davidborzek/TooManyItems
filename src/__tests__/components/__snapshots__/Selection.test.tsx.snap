// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection renders correctly with full props and opened 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "flex": 1,
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "white",
            "height": "50%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              }
            }
          >
            Some title
          </Text>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text />
          </View>
        </View>
        <RCTScrollView>
          <View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "borderBottomColor": "#e1e1e1",
                    "borderBottomWidth": 1,
                    "padding": 16,
                  }
                }
              >
                Some label
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#86939e",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Some label
      </Text>
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          autoCompleteType=""
          editable={false}
          enablesReturnKeyAutomatically={true}
          placeholder="Some placeholder"
          placeholderTextColor="#86939e"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Selection renders correctly with required props, no items and closed 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "flex": 1,
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "white",
            "height": "50%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              }
            }
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text />
          </View>
        </View>
        <RCTScrollView>
          <View />
        </RCTScrollView>
      </View>
    </View>
  </Modal>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          autoCompleteType=""
          editable={false}
          enablesReturnKeyAutomatically={true}
          placeholderTextColor="#86939e"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Selection renders correctly with required props, with items and closed 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "flex": 1,
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "white",
            "height": "50%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              }
            }
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text />
          </View>
        </View>
        <RCTScrollView>
          <View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "borderBottomColor": "#e1e1e1",
                    "borderBottomWidth": 1,
                    "padding": 16,
                  }
                }
              >
                Some label
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          autoCompleteType=""
          editable={false}
          enablesReturnKeyAutomatically={true}
          placeholderTextColor="#86939e"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`Selection renders correctly with required props, with items and opened 1`] = `
<View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    statusBarTranslucent={true}
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.5)",
          "flex": 1,
          "justifyContent": "flex-end",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "white",
            "height": "50%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 16,
                "fontWeight": "600",
              }
            }
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text />
          </View>
        </View>
        <RCTScrollView>
          <View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Object {
                    "borderBottomColor": "#e1e1e1",
                    "borderBottomWidth": 1,
                    "padding": 16,
                  }
                }
              >
                Some label
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
  </Modal>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          autoCompleteType=""
          editable={false}
          enablesReturnKeyAutomatically={true}
          placeholderTextColor="#86939e"
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
  </View>
</View>
`;
